### ISSUE TITLE
Mean reward is not increasing

### ISSUE THREAD

**Issue by MrGitGo** [Unity-Technologies/ml-agents#2061]
![issueBild](https://user-images.githubusercontent.com/33628476/57978605-44464780-7a11-11e9-8fe6-617f67dae3a6.png)

- Set-up: 5 vs 5 shooter game, where the agents can move in all directions and shoot
- Goal: The goal of a team is to eliminate the other
- Agents: All 10 Agents are linked to one brain
- Agent Reward Function:
    - This applies for both teams:
    - -0.2 when agent dies
    - 0.5 when agent kills
    - -1.0 when all agents in team dead
    - 1.0 when all opponents are dead
    - after a kill it checks if there is an oppononent remaining, if yes: -0.3 for the team
(There is one thing to say: these rewards are always given to all agents in the team)

-Brain: One Brain with the following action space:
    - Vector action space: (Discrete) 6 possible actions (shoot,move left, -right, rotate, ...)

Vector Observation:

```
if (useVectorObs)
        {
            var rayDistance = 8f;
            var rayDistance2 = 30f;
            float[] rayAngles = { 0f, 90f, 135f, 180f, 70f };
            float[] rayAngles2 = {45f, 90f,110f};
            var detectableObjects = new[] { "agent", "wall","counterAgent" };
            Vector3 forward = transform.forward;
            forward.y = 0;

            AddVectorObs(Quaternion.LookRotation(forward).eulerAngles.y);

            AddVectorObs(rayPer.Perceive(rayDistance2, rayAngles2, detectableObjects, 2.8f, 0f));

            AddVectorObs(rayPer.Perceive(rayDistance, rayAngles, detectableObjects, 1.5f, 0f));

            AddVectorObs(gameObject.GetComponent<Rigidbody>().velocity);
        }
```

Tensorboard:

![tensorboard4mio](https://user-images.githubusercontent.com/33628476/57978830-b4a29800-7a14-11e9-963d-9af40d27e963.png)

And I also changed the arena after every million steps:

![progressTo4Mio](https://user-images.githubusercontent.com/33628476/57978842-d734b100-7a14-11e9-9931-35839f6b59e3.png)

And the max steps is  (in Academy): 2000  
num_layers: 1
normalize: true
batch_size: 64
buffer_size: 10240
hidden_units: 128 

Could you give me hints in solving the problem with the rewards? I have read here somewhere in the issues that I should be very sparsly with negative rewards, but still I want to ask you, too. I hope you can help me

**Comment by mattinjersey** [comment]
there are a lot of hyperparameters including gamma, lambda, learningrate, number of layers,...
maybe by adjusting these it will work.
also you could try training longer for days...

reinforcement learning can be frustrating business I guess!

**Comment by xiaomaogy** [comment]
I think you should
1) Make the scene simpler by changing it to 1 vs 1 instead of 5 vs 5, then iterate from there. 
2) Use the default hyperparameter first before trying your own. 

@MrGitGo

**Comment by MrGitGo** [comment]
Ok, I will try this out. But what is the point of trying it with 1 vs 1 first. How does it influence the training? I thought it would be better if I have more agents.
@xiaomaogy

**Comment by xiaomaogy** [comment]
No, more agent will make the training more complicated and harder to debug. You should start with the simplest possible scene that should absolutely work, then iterate from there. @MrGitGo

**Comment by MrGitGo** [comment]
So I put the agents incrementally in the game (1st phase: 1 vs 1, 2nd 2 vs 2, ..). I did this until 3 vs. 3 and I was at step 9 million. After a while I made the observation, that I am getting mean rewards: 0.0 and std rewards: 0.0. And that the whole time. When I changed the spawnpoints I had again sometimes positive rewards and sometimes negative. I realised when the agents avoid each other and don't kill anyone they get the 0.0 reward. Maybe I am doing it wrong with the 3 vs. 3 agents.

My main goal is, after I successfully trained the agents, to play with one character vs. 3 other agents. So it is 1 human vs. 3 agents. My question is, is it correct to train like that? Should I put 3 agents and one character that is randomly moving and shooting? I could put the agent also one after another.

The reason why I have 3 vs 3 agents is, that I thought, that it would be better for the training. The more agents I have the better and faster they would train. But I failed at this I guess. So how and what should I do next to reach my goal? The last thing my agents were doing was to avoid each other until I changed the spawnpoints. But I guess after a while they will avoid each other again. I guess it is because of the rewards: when one agent is killing an other, the one will get positive reward and the other negative. And both are connected to one brain. So in order to get no negative reward, they try to avoid each other.


Edit: Everytime I increse the number of the agents, it seems that the training looks ok, but you see yourself here. I have a one minute video of the game and some pics:

VIdeo:
https://www.dropbox.com/s/fofsxe7lj34kk4w/issues.mp4?dl=0

Tensorboard:

![issues2](https://user-images.githubusercontent.com/33628476/58536049-c09a1100-81ef-11e9-90ed-34f8ccc4a5ea.PNG)


And I am using the RayPerception script, but it seems that it is not fast enough for my training. Look here:

![issues](https://user-images.githubusercontent.com/33628476/58536095-e0313980-81ef-11e9-86e3-c3e0d267e847.png)


The lines are normally attached to the player (0,0,0), but if they move wildly, this happens.

And One more thing to ask: I see on my tensorboard, that the entropy is not decreasing fast enough. Can I decrease the value of beta and continue with my training? Or do I have to train from the beginning, when I change a parameter? Should I change a parameter, when yes which one, to which value? 

This is the link to my .yaml: https://www.dropbox.com/s/4src7kt2ykwfqu2/trainer_config.yaml?dl=0

I am using the default

**Comment by mattinjersey** [comment]
I don't think you should give -0.3 if there is an opponent still alive? Because it looks like the one agent will get -0.2 and the other agent will get 0.2 so that the rewards are equal and the system will find a solution that they avoid each other. So I think the reward structure should be modified so its more reward for the hunter.

**Comment by xiaomaogy** [comment]
@MrGitGo 
A few things you can try. 

1. Remove the negative reward for dying. The positive reward should already be enough. 
2. Make your training scene much smaller. (like a really small room). After your agent learns to kill each other, you can iterate on that model (manually use curriculum learning)
3. Start with 1 vs 1 agent. 
4. Have as many 1 vs 1 training scene as possible to speed up the training.

**Comment by MrGitGo** [comment]
1. I guess I have to restart my model and begin from step 0 again

2. Ok, I will start with as many 1 vs 1 secenes as possible, but how can I iterate the agents then, when they are already active in the scene?

**Comment by xiaomaogy** [comment]
What do you mean by "how can I iterate the agents when they are already active in the scene"? I guess you mean how do you reset the agents so that they are not stuck after winning? You just implement the corresponding AgentReset method or the AcademyReset to reset all of the Agents periodically. @MrGitGo

**Comment by MrGitGo** [comment]
Ok, I am just calling the agent done() method and it works

**Comment by unityjeffrey** [comment]
Thank you for the discussion. We are closing this issue due to inactivity. Feel free to reopen it if you’d like to continue the discussion though.
